packages/core/src/modules/oob/OutOfBandModule.ts

```diff
@@ -0,0 +1,181 @@
+import type { AgentMessage } from '../../agent/AgentMessage'
+import type { AgentMessageReceivedEvent } from '../../agent/Events'
+import type { ConnectionRecord } from '../../modules/connections'
+import type { OutOfBandMessageOptions } from './OutOfBandMessage'
+
+import { Lifecycle, scoped } from 'tsyringe'
+
+import { EventEmitter } from '../../agent/EventEmitter'
+import { AgentEventTypes } from '../../agent/Events'
+import { MessageSender } from '../../agent/MessageSender'
+import { createOutboundMessage } from '../../agent/helpers'
+import { AriesFrameworkError } from '../../error'
+import { ConnectionService, ConnectionInvitationMessage } from '../connections'
+import { DiscoverFeaturesQueryMessage, DiscoverFeaturesService } from '../discover-features'
+import { MediationRecipientService } from '../routing'
+
+import { OutOfBandMessage } from './OutOfBandMessage'
+
+@scoped(Lifecycle.ContainerScoped)
+export class OutOfBandModule {
+  private connectionService: ConnectionService
+  private mediationRecipientService: MediationRecipientService
+  private disoverFeaturesService: DiscoverFeaturesService
+  private messageSender: MessageSender
+  private eventEmitter: EventEmitter
+
+  public constructor(
+    connectionService: ConnectionService,
+    mediationRecipientService: MediationRecipientService,
+    disoverFeaturesService: DiscoverFeaturesService,
+    messageSender: MessageSender,
+    eventEmitter: EventEmitter
+  ) {
+    this.connectionService = connectionService
+    this.mediationRecipientService = mediationRecipientService
+    this.disoverFeaturesService = disoverFeaturesService
+    this.messageSender = messageSender
+    this.eventEmitter = eventEmitter
+  }
+
+  /**
+   * Creates new connection record and use its keys for out-of-band message that works as a connection invitation.
+   * It uses discover features to find out what handshake protocols the agent supports.
+   *
+   * @param options Optinal attributes contained in out-of-band message
+   * @returns Created connection record and out-of-band message
+   */
+  public async createInvitation(
+    options: OutOfBandMessageOptions
+  ): Promise<{ outOfBandMessage: OutOfBandMessage; connectionRecord: ConnectionRecord }> {
+    // Discover what handshake protocols are supported
+    const queryMessage = new DiscoverFeaturesQueryMessage({
+      query: `*`,
+    })
+    const featuresMessage = await this.disoverFeaturesService.createDisclose(queryMessage)
+    const { protocols } = featuresMessage
+
+    const handshakeProtocols = ['https://didcomm.org/didexchange', 'https://didcomm.org/connections']
+
+    const supportedHandshakeProtocols = protocols
+      .map((p) => p.protocolId.slice(0, -1))
+      .filter((pId) => handshakeProtocols.find((hp) => pId.startsWith(hp)))
```

Comment:
I think it would be nice to extract this to the discover features service and call it like this:

``` ts
const supported = await this.disoverFeaturesService.getSupportedProtocols(['https://didcomm.org/didexchange', 'https://didcomm.org/connections'])
```

Reply:
That's a good point I also felt like it should be extracted somewhere. I moved it to `disoverFeaturesService`. Eventually, it can be part of the dispatcher.

Reply:
Updated.